### PR TITLE
Created the Puzzle.py module with the Puzzle, SlidingPuzzle, PuzzlePiece, and ExamplePuzzle classes

### PR DESCRIPTION
**Pull request by Inventor4life**
an example_main.py file has been created to show how to use the new module. I will be updating the documentation for the Puzzle class at a later time
requirements.txt has been updated with the required dependencies.

### CODE REVIEW

File: Puzzle.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
+import numpy as np
+import pygame
+import PIL
+import random
+
+'''
```
Comment (A5rocks) [inline Puzzle.py]: docstrings normally go in the class (and `"""` instead of `'''`)

File: Puzzle.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+Parent class Puzzle
+    Description:
+        All image puzzles are a subset of this class.
+        This class splits a puzzle image up into individual pieces that can be modified, moved, and deleted.
+        It is generally assumed that the puzzle is square, so rectangular puzzles will have rectangular pieces.
```
Comment (A5rocks) [inline Puzzle.py]: I don't think this assumption is necessary for doing things (though I haven't looked at the code yet...)

File: Puzzle.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+            this is used to determine which piece was clicked when
+            the puzzle is not in the top-left corner of the window.
+
+    Methods:
+
+        modify_image(image, output_size)
+            Resizes the input image to the output size.
+            Splits the image into total_pieces parts, then creates a PuzzlePiece object from the image.
+            PuzzlePiece object is stored in a list called pieces.
+            Raises an Exception if the resized image cannot be divided into total_pieces without remainder.
```
Comment (A5rocks) [inline Puzzle.py]: These method docstrings should go within the method they're describing, I think.

File: Puzzle.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        swapping the top two pieces:
+        orderlist is [1, 0, 2, 3]
+        puzzle looks like:
+        [ 1 , 0 ,
+          2 , 3 ]
+'''
+
+
+class Puzzle:
+    def __init__(self, image, pieces_per_side, output_size, puzzle_pos):
```
Comment (A5rocks) [inline Puzzle.py]: I notice a lack of type hints! You can probably do:

```suggestion
    def __init__(self, image: PIL.Image, pieces_per_side: int, output_size: int, puzzle_pos: tuple[int, int]):
```

I think! Just make sure mypy is fine with that. I think you can probably figure out the methods' type hints!

File: Puzzle.py
Comment (A5rocks) [inline Puzzle.py:1]: python conventions: lowercase file names

File: example_main.py
Comment (A5rocks) [inline example_main.py:1]: `example_main.py` -> `main.py`?

File: example_main.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+Create new puzzle object given puzzle pieces
+'''
+sliding_puzzle = Puzzle.SlidingPuzzle(puzzle_image, PIECES_PER_SIDE, (380, 500))
+
+running = True
+mouse_held = False
+mouse_was_held = False
+screen.fill((255, 0, 0))
+screen.blit(sliding_puzzle.image, (0, 0))
+while running:
```
Comment (A5rocks) [inline example_main.py]: Mind putting this code behind an `if __name__ == "__main__"`?

File: sample_images/Monalisa.png
Comment (A5rocks) [inline sample_images/Monalisa.png:1]: I'm not completely sure about storing images in the git repository yet as (due to how git works) that means every single clone for the git repo will include every single image (all however many bytes there are) whether they are deleted or not.

... But I'm not sure if there's a nice way to include them.

**Review by A5rocks — CHANGES_REQUESTED** [review]
Very surface-level review. I haven't made my own puzzle so I'm not sure if your provided interface is good enough for a range of puzzles. Maybe others can chime in with reviews!

File: Puzzle.py (hunk 2 above)
Comment (Inventor4life) [reply Puzzle.py]: I changed it. I was referring to the number of pieces wide vs the number of pieces tall.

**Pull request closed without merging**